<a href="https://colab.research.google.com/github/Mridener/IN498-Capstone-in-Analytics/blob/main/IN498_MakaylaRidener_Unit2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# GitHub Repository

A GitHub repository was used to track project progress and store all assignment files.

In [35]:
%cd /content/IN498-Capstone-in-Analytics

!git add .
!git commit -m "Completed data curation of final dataset"
!git push

/content/IN498-Capstone-in-Analytics
On branch main
Your branch is up to date with 'origin/main'.

nothing to commit, working tree clean
Everything up-to-date


#Parsing CSV Files

Each raw file was read and converted into a CSV file. The script first attempted to read the file using UTF-8 encoding and then used UTF-16 when necessary. Column headers were added, quotation marks were removed, and each row was split into separate columns using commas as delimiters. The parsed files were then saved in UTF-8 format with each data value stored in its own cell.

In [22]:
import csv

#Files to parse
files = [
    "/content/IN498-Capstone-in-Analytics/retained_installers_com.foo.bar_201904_country.csv",
    "/content/IN498-Capstone-in-Analytics/retained_installers_com.foo.bar_201905_country.csv",
    "/content/IN498-Capstone-in-Analytics/retained_installers_com.foo.bar_201906_country.csv",
    "/content/IN498-Capstone-in-Analytics/retained_installers_com.foo.bar_201907_country.csv",
    "/content/IN498-Capstone-in-Analytics/retained_installers_com.foo.bar_201908_country.csv",
    "/content/IN498-Capstone-in-Analytics/retained_installers_com.foo.bar_201909_country.csv",
    "/content/IN498-Capstone-in-Analytics/retained_installers_com.foo.bar_201910_country.csv",
    "/content/IN498-Capstone-in-Analytics/retained_installers_com.foo.bar_201911_country.csv",
    "/content/IN498-Capstone-in-Analytics/retained_installers_com.foo.bar_201912_country.csv",]

#Column names
columns = [
    "Date",
    "Package_Name",
    "Country",
    "Store_Listing_Visitors",
    "Installers",
    "Visitor-to-Installer_conversion_rate",
    "Installers_retained_for_1_day",
    "Installer-to-1_day_retention_rate",
    "Installers_retained_for_7_days",
    "Installer-to-7_days_retention_rate",
    "Installers_retained_for_15_days",
    "Installer-to-15_days_retention_rate",
    "Installers_retained_for_30_days",
    "Installer-to-30_days_retention_rate"]

#Parse each raw file
for file in files:

  try:
    with open(file, "r", encoding="utf-8-sig") as raw_file:
      lines = raw_file.readlines()

  except UnicodeDecodeError:
    with open(file, "r", encoding="utf-16") as raw_file:
      lines = raw_file.readlines()

  parsed_file = file.replace(".csv", "_parsed.csv")

  #Write parsed CSV file
  with open(parsed_file, "w", newline="", encoding="utf-8") as output_file:

    writer = csv.writer(
        output_file,
        delimiter=",",
        quoting=csv.QUOTE_MINIMAL
      )

    writer.writerow(columns)

    #Parse each row
    for line in lines:
      line = line.strip()
      if line != "":

        #Remove quotes within row
        line = line.strip('"')

        #Split row into columns
        row = line.split(",")

        #Remove quotes from field
        row = [value.strip('"') for value in row]

        writer.writerow(row)

  print(f"Created: {parsed_file}")

print("All nine files were parsed and saved as UTF-8 csv files.")

Created: /content/IN498-Capstone-in-Analytics/retained_installers_com.foo.bar_201904_country_parsed.csv
Created: /content/IN498-Capstone-in-Analytics/retained_installers_com.foo.bar_201905_country_parsed.csv
Created: /content/IN498-Capstone-in-Analytics/retained_installers_com.foo.bar_201906_country_parsed.csv
Created: /content/IN498-Capstone-in-Analytics/retained_installers_com.foo.bar_201907_country_parsed.csv
Created: /content/IN498-Capstone-in-Analytics/retained_installers_com.foo.bar_201908_country_parsed.csv
Created: /content/IN498-Capstone-in-Analytics/retained_installers_com.foo.bar_201909_country_parsed.csv
Created: /content/IN498-Capstone-in-Analytics/retained_installers_com.foo.bar_201910_country_parsed.csv
Created: /content/IN498-Capstone-in-Analytics/retained_installers_com.foo.bar_201911_country_parsed.csv
Created: /content/IN498-Capstone-in-Analytics/retained_installers_com.foo.bar_201912_country_parsed.csv
All nine files were parsed and saved as UTF-8 csv files.


# Explorer

Each parsed CSV file was loaded into a data frame and reviewed using pandas. The first 10 rows, shape, description, counts, mean, and mode were generated for each file. This process provided an overview of the data before merging the files.

In [25]:
import pandas as pd
import sys

#Ignoring warnings
if not sys.warnoptions:
  import warnings
  warnings.simplefilter("ignore")

#Set the writeFunction boolean
#True = write to console
#False = write to file for assignment turnin
PRINT =  False

#Open a file handle for assignment results.
if PRINT == False:
  f = open("/content/IN498-Capstone-in-Analytics/IN498_Explorer.txt", "w")

###############################################
##
##PURPOSE: Write to console or file based on
##  the writeFunction variable
##      True = write to concole
##      False = write to file for assignment turnin
##
##INPUT: Message to write (message1)
##   Optional: message2
##
##OUTPUT: None
##
###############################################
def writeFunction(message1, *message2):

#Print to console
  if PRINT:
    print(message1)
    print(message2)
    print()
#Print to file for assignment
  else:
    f.write(str(message1))
    f.write(str(message2))
    f.write("\n\n")


#Widen the column display
pd.set_option('max_colwidth',500)

#Files to analyze in a list
files = ["/content/IN498-Capstone-in-Analytics/retained_installers_com.foo.bar_201904_country_parsed.csv",
         "/content/IN498-Capstone-in-Analytics/retained_installers_com.foo.bar_201905_country_parsed.csv",
         "/content/IN498-Capstone-in-Analytics/retained_installers_com.foo.bar_201906_country_parsed.csv",
         "/content/IN498-Capstone-in-Analytics/retained_installers_com.foo.bar_201907_country_parsed.csv",
         "/content/IN498-Capstone-in-Analytics/retained_installers_com.foo.bar_201908_country_parsed.csv",
         "/content/IN498-Capstone-in-Analytics/retained_installers_com.foo.bar_201909_country_parsed.csv",
         "/content/IN498-Capstone-in-Analytics/retained_installers_com.foo.bar_201910_country_parsed.csv",
         "/content/IN498-Capstone-in-Analytics/retained_installers_com.foo.bar_201911_country_parsed.csv",
         "/content/IN498-Capstone-in-Analytics/retained_installers_com.foo.bar_201912_country_parsed.csv"]

#Read data into a DataFrame using these columns
##"Date","Package_Name","Country","Store_Listing_Visitors",
##"Installers","Visitor-to-Installer_conversion_rate",
##"Installers_retained_for_1_day","Installer-to-1_day_retention_rate",
##"Installers_retained_for_7_days","Installer-to-7_days_retention_rate",
##"Installers_retained_for_15_days","Installer-to-15_days_retention_rate",
##"Installers_retained_for_30_days","Installer-to-30_days_retention_rate"

col_names = ["Date",
             "Package_Name",
             "Country",
             "Store_Listing_Visitors",
             "Installers",
             "Visitor-to-Installer_conversion_rate",
             "Installers_retained_for_1_day",
             "Installer-to-1_day_retention_rate",
             "Installers_retained_for_7_days",
             "Installer-to-7_days_retention_rate",
             "Installers_retained_for_15_days",
             "Installer-to-15_days_retention_rate",
             "Installers_retained_for_30_days",
             "Installer-to-30_days_retention_rate"
]

#Retention columns for mean and mode
retention_columns = [
    "Installers_retained_for_1_day",
    "Installers_retained_for_7_days",
    "Installers_retained_for_15_days",
    "Installers_retained_for_30_days"
]

#Parse files to analyze structure of CSV files and data
for file in files:

  #Print the file
  writeFunction("#########File#########", file)

  #Get the data frame from the file
  df = pd.read_csv(file)

  ########### EXPLORE THE DATA SET #######################

  #Print the top 10 rows
  writeFunction("Top 10 Rows")
  writeFunction(df.head(10))

  #Print the shape
  writeFunction("Shape")
  writeFunction(df.shape)

  #Print the description of the entire data set
  writeFunction("Description")
  writeFunction(df.describe(include="all"))

  #Print the counts of the data
  writeFunction("Counts")
  writeFunction(df.count())

# Get mean and mode for installers retained 1, 7, 15, and 30 days
  for col in retention_columns:
    df[col] = pd.to_numeric(df[col], errors="coerce")

  for col in retention_columns:
    writeFunction(f"Mean for {col}")
    writeFunction(df[col].mean())

    writeFunction(f"Mode for {col}")
    writeFunction(df[col].mode())

#Close the file handle
if PRINT == False:
      f.close()

**Shape of Each Data Frame**
* April 2019: (889, 14)
* May 2019: (895, 14)
* May 2019: (895, 14)
* June 2019: (790, 14)
* July 2019: (727, 14)
* August 2019: (778, 14)
* September 2019: (560, 14)
* October 2019: (582, 14)
* November 2019: (595, 14)
* December 2019: (601, 14)

The data frames ranged from 560 to 895 rows and contained 14 columns each. Missing values were found mainly in the Date and Country columns, with some missing values in the May and July 30-day retention columns. The mean values ranged from 1.36 to 1.81 for 1-day retention, 1.10 to 1.39 for 7-day retention, 0.95 to 1.20 for 15-day retention, and 0.79 to 1.03 for 30-day retention. The mode for all retention columns was 0. Based on the analysis, the datasets were mostly complete, retention values decreased as the retention period increased, and many records had no retained installers. Overall, the data appeared consistent and ready for merging and further analysis.

# File Merging

The nine parsed CSV files were read into data frames and stored in a list. The data frames were then merged into a single data frame using the pandas concat() function. The merged data frame was saved as retained_installers_final.csv.

In [28]:
import pandas as pd

#Parsed files
files = [
    "/content/IN498-Capstone-in-Analytics/retained_installers_com.foo.bar_201904_country_parsed.csv",
    "/content/IN498-Capstone-in-Analytics/retained_installers_com.foo.bar_201905_country_parsed.csv",
    "/content/IN498-Capstone-in-Analytics/retained_installers_com.foo.bar_201906_country_parsed.csv",
    "/content/IN498-Capstone-in-Analytics/retained_installers_com.foo.bar_201907_country_parsed.csv",
    "/content/IN498-Capstone-in-Analytics/retained_installers_com.foo.bar_201908_country_parsed.csv",
    "/content/IN498-Capstone-in-Analytics/retained_installers_com.foo.bar_201909_country_parsed.csv",
    "/content/IN498-Capstone-in-Analytics/retained_installers_com.foo.bar_201910_country_parsed.csv",
    "/content/IN498-Capstone-in-Analytics/retained_installers_com.foo.bar_201911_country_parsed.csv",
    "/content/IN498-Capstone-in-Analytics/retained_installers_com.foo.bar_201912_country_parsed.csv"]

#Store data frames
dataframes = []

#Read parsed files
for file in files:
  df = pd.read_csv(file)
  dataframes.append(df)

#Merge data frames
final_df = pd.concat(dataframes, ignore_index=True)

#Save final data frame
final_df.to_csv("retained_installers_final.csv", index=False)

print("Final merged file created.")
print(final_df.shape)


Final merged file created.
(6417, 14)


# Final Explorer

The final merged dataset was loaded into a data frame and reviewed using pandas. The first 10 rows, shape, description, counts, mean, and mode were generated for the final dataset. This process provided a better understanding of the merged data and helped identify missing values before data curation.

In [31]:
from os import write
import pandas as pd
import sys

#Ignoring warnings
if not sys.warnoptions:
  import warnings
  warnings.simplefilter("ignore")

#Set the writeFunction boolean
#True = write to console
#False = write ot file for assignment turnin
PRINT = False

#Open a file handle for assignment results.
if PRINT == False:
  f = open("/content/IN498-Capstone-in-Analytics/IN498_Final_Explorer.txt", "w")
###############################################
##
##PURPOSE: Write to console or file based on
##  the writeFunction variable
##      True = write to concole
##      False = write to file for assignment turnin
##
##INPUT: Message to write (message1)
##   Optional: message2
##
##OUTPUT: None
##
###############################################
def writeFunction(message1, *message2):

  #Print to console
  if PRINT:
    print(message1)
    print(message2)
    print()
  #Print to file for assignment
  else:
    f.write(str(message1))
    f.write(str(message2))
    f.write("\n\n")

#Widen the column display
pd.set_option('max_colwidth',500)

#Read data into a DataFrame using these columns
##"Date","Package_Name","Country","Store_Listing_Visitors",
##"Installers","Visitor-to-Installer_conversion_rate",
##"Installers_retained_for_1_day","Installer-to-1_day_retention_rate",
##"Installers_retained_for_7_days","Installer-to-7_days_retention_rate",
##"Installers_retained_for_15_days","Installer-to-15_days_retention_rate",
##"Installers_retained_for_30_days","Installer-to-30_days_retention_rate"
col_names = ["Date",
             "Package_Name",
             "Country",
             "Store_Listing_Visitors",
             "Installers",
             "Visitor-to-Installer_conversion_rate",
             "Installers_retained_for_1_day",
             "Installer-to-1_day_retention_rate",
             "Installers_retained_for_7_days",
             "Installer-to-7_days_retention_rate",
             "Installers_retained_for_15_days",
             "Installer-to-15_days_retention_rate",
             "Installers_retained_for_30_days",
             "Installer-to-30_days_retention_rate"]

#Parse files to analyze structure of CSV files and data
#Get the data frame from the file
############## YOUR FILENAME MAY DIFFER #####################
df = pd.read_csv("/content/IN498-Capstone-in-Analytics/retained_installers_final.csv")

########### EXPLORE THE DATA SET #######################

#Print the top 10 rows
writeFunction("Top 10 Rows")
writeFunction(df.head(10))

#Print the shape
writeFunction("Shape")
writeFunction(df.shape)

#Print the description of the entire data set
writeFunction("Description")
writeFunction(df.describe(include="all"))

#Print the counts of the data
writeFunction("Counts")
writeFunction(df.count())

#Get the mean and mode for installers retained 1, 7, 15, and 30 days
retention_columns = ["Installers_retained_for_1_day",
                     "Installers_retained_for_7_days",
                     "Installers_retained_for_15_days",
                     "Installers_retained_for_30_days"]

for col in retention_columns:
  df[col] = pd.to_numeric(df[col], errors="coerce")

for col in retention_columns:
  writeFunction(f"Mean for {col}")
  writeFunction(df[col].mean())

  writeFunction(f"Mode for {col}")
  writeFunction(df[col].mode())

#Close the file handle
if PRINT == False:
  f.close()

The final data frame contained 6,417 rows and 14 columns. The counts showed 6,403 values in the Date column, 6,358 values in the Country column, 6,409 values in the Installers_retained_for_30_days column, and 6,409 values in the Installer-to-30_days_retention_rate column. All other columns contained 6,417 values. The mean values for the retention columns were 1.52 for 1-day retention, 1.19 for 7-day retention, 1.03 for 15-day retention, and 0.87 for 30-day retention. The mode for all four retention columns was 0, indicating that zero retained installers was the most common value in the final dataset.

# Data Curation

The merged dataset was loaded into a data frame and copied for cleaning. The Date column was standardized using a consistent date format, and missing values in the Date and Country columns were replaced with "Unknown." Rows containing missing 30-day retention data were removed from the dataset. The cleaned data was then saved as a new CSV file for further analysis.

In [33]:
import pandas as pd

#Read merged dataset
df = pd.read_csv("/content/IN498-Capstone-in-Analytics/retained_installers_final.csv")

#Create copy for cleaning
cleaned_df = df.copy()

#Standardize Date format
cleaned_df["Date"] = pd.to_datetime(cleaned_df["Date"],
                                    errors="coerce",
                                    format="mixed")

cleaned_df["Date"] = cleaned_df["Date"].fillna("Unknown")

#Replace missing Country values
cleaned_df["Country"] = cleaned_df["Country"].fillna("Unknown")

# Remove rows with missing 30-day retention data
cleaned_df = cleaned_df.dropna(subset=["Installers_retained_for_30_days",
                                       "Installer-to-30_days_retention_rate"])

# Save cleaned version as a new file
cleaned_df.to_csv("/content/IN498-Capstone-in-Analytics/retained_installers_final_cleaned.csv",
                  index=False)

print("Cleaned file created.")
print(cleaned_df.shape)

Cleaned file created.
(6409, 14)


The missing values in the Date and Country columns were replaced with "Unknown" so those records could remain in the dataset. Since those columns contain descriptive information, replacing the missing values did not affect the remaining data. The missing values in the 30-day retention columns were handled by removing those rows because those columns contain numerical retention data. Replacing the missing retention values with "Unknown" would have mixed text with numerical values and could have created problems for future calculations. Since only a small number of records contained missing 30-day retention values, removing those rows had little impact on the overall dataset. This approach allowed most of the data to be preserved while creating a cleaner final dataset.